### ISSUE TITLE
Vim debug mode could not work

### ISSUE THREAD

**Issue by hiberabyss** [google/vim-searchindex#14]
## Problem
Because of the following `cmap` in this plugin, the vim debug mode could not work in some case.

    cmap <expr> <CR> <SID>handle_cr()

If remove this map, there won't be such problem. 

## How to reproduce
Execute `:debug 1` and then press `q`, the debug mode could not be quitted. Following message will be shown:
![image](https://cloud.githubusercontent.com/assets/2898670/24835791/f1c7fbd4-1d3d-11e7-83d2-2b69932dc174.png)

**Comment by rburny** [comment]
I looked into that, but I don't see a way to fix it.

The reason why this mapping is necessary is that it is (to my knowledge) the only way to detect when the search is performed. There's no `SearchDone` autocmd, and there's not even `CmdEnter` autocmd that would allow setting this mapping condionally.

As a workaround, you can use `Ctrl-J` in debug mode - it's equivalent to `<CR>`, but not mapped. Let me know if you have an idea for a better fix.

**Comment by lacygoill** [comment]
@hiberabyss 

I think that any mapping defined in command-line mode may interfere when you're in debug mode. But, IIRC, only if the `rhs` contains the evaluation of a custom function.

FWIW, I've defined a command to toggle on/off all mappings defined in command-line mode (`:ToggleEditingCommands`). I never use it manually, but I call it inside other custom functions.

Here's some code, using this command, which I think should get around the issue you describe:

    cnorea <expr> debug getcmdtype() == ':' && getcmdline() =~# '\v^((tab<bar>vert[ical])\s+)?debug$' ? 'Debug' : 'debug'

    com! -bar -nargs=1 Debug exe s:debug(<q-args>)
    com! -bar -nargs=1 ToggleEditingCommands exe s:toggle_editing_commands(<args>)

    fu! s:debug(cmd) abort
        try
            ToggleEditingCommands 0
            exe 'debug '.a:cmd
        catch
            return 'echoerr '.string(v:exception)
        finally
            ToggleEditingCommands 1
        endtry
        return ''
    endfu

    fu! s:toggle_editing_commands(enable) abort
        try
            if a:enable
                call s:tmp_mappings_restore(get(s:, 'my_editing_commands', []))
            else
                let lhs_list = map(split(execute('cno'), '\n'), 'matchstr(v:val, ''\vc\s+\zs\S+'')')
                call filter(lhs_list, '!empty(v:val)')
                let s:my_editing_commands = s:tmp_mappings_save(lhs_list, 'c', 1)
                for lhs in lhs_list
                    exe 'cunmap '.lhs
                endfor
            endif
        catch
            return 'echoerr '.string(v:exception)
        endtry
    endfu

    fu! s:tmp_mappings_save(keys, mode, global) abort
        let mappings = {}
        for l:key in a:keys
            let buf_local_map = maparg(l:key, a:mode, 0, 1)

            sil! exe a:mode.'unmap <buffer> '.l:key

            let map_info        = maparg(l:key, a:mode, 0, 1)
            let mappings[l:key] =   !empty(map_info)
            \                     ?     map_info
            \                     :     {
            \                             'unmapped' : 1,
            \                             'buffer'   : 0,
            \                             'lhs'      : l:key,
            \                             'mode'     : a:mode,
            \                           }
            call s:tmp_mappings_restore({l:key : buf_local_map})
        endfor
        return mappings
    endfu

    fu! s:tmp_mappings_restore(mappings) abort
        if empty(a:mappings)
            return
        endif
        for mapping in values(a:mappings)
            if !has_key(mapping, 'unmapped') && !empty(mapping)
                exe     mapping.mode
                \ . (mapping.noremap ? 'noremap   ' : 'map ')
                \ . (mapping.buffer  ? ' <buffer> ' : '')
                \ . (mapping.expr    ? ' <expr>   ' : '')
                \ . (mapping.nowait  ? ' <nowait> ' : '')
                \ . (mapping.silent  ? ' <silent> ' : '')
                \ .  mapping.lhs
                \ . ' '
                \ . substitute(mapping.rhs, '<SID>', '<SNR>'.mapping.sid.'_', 'g')
            elseif has_key(mapping, 'unmapped')
                sil! exe mapping.mode.'unmap '
                                    \ .(mapping.buffer ? ' <buffer> ' : '')
                                    \ . mapping.lhs
            endif
        endfor
    endfu

It installs an abbreviation, which automatically replaces `:debug` with `:Debug`. The latter is a custom command used as a wrapper around `:debug` to temporarily disable all your mappings in command-line mode, including `<CR>`. They should be restored once you leave the debug mode.

@rburny 
In case it helps you in the future, recently ([8.0.1206](https://github.com/vim/vim/commit/fafcf0dd59fd9c4ef743bb333ae40d1d322b6079)), Vim has introduced 2 new events called `CommandlineEnter` and `CommandlineLeave`. Here is a [gist](https://gist.github.com/lacygoill/88437bcbbe90d1d2e2787695e1c1c6a9) about it.

**Comment by rburny** [comment]
Thanks for the pointers, @lacygoill. In the end, I figured out the problem is caused by the helper method `handle_cr`. Since it is written in Vimscript, the debugger tries to debug it in a recursive fashion. After inlining the method, everything works well.

`CmdlineEnter` could indeed be used to implement the mapping in a slightly less intrusive way. But given that many users may still run Vim 7, I'd have to support both implementations. Therefore, I decided to just keep the old one - but dropped the note to use `CmdlineEnter` should we see any more issues.
